src/devices/machine/nandflash.h

```diff
@@ -0,0 +1,150 @@
+// license:BSD-3-Clause
+// copyright-holders:Raphael Nabet
+
+#ifndef MAME_MACHINE_NANDFLASH_H
+#define MAME_MACHINE_NANDFLASH_H
+
+#pragma once
+
+/***************************************************************************
+    TYPE DEFINITIONS
+***************************************************************************/
+
+// ======================> nand_device
+class nand_device : public device_t, public device_nvram_interface
+{
+public:
+	// construction/destruction
+	nand_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// device-level overrides
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	virtual void device_resolve_objects() override;
+
+	// device_nvram_interface overrides
+	virtual void nvram_default() override;
+	virtual bool nvram_read(util::read_stream &file) override;
+	virtual bool nvram_write(util::write_stream &file) override;
+
+	auto rnb_wr_callback() { return m_write_rnb.bind(); }
+
+	int is_present();
+	int is_protected();
+	int is_busy();
+
+	uint8_t data_r();
+	void command_w(uint8_t data);
+	void address_w(uint8_t data);
+	void data_w(uint8_t data);
+
+protected:
+	enum sm_mode_t
+	{
+		SM_M_INIT,       // initial state
+		SM_M_READ,       // read page data
+		SM_M_PROGRAM,    // program page data
+		SM_M_ERASE,      // erase block data
+		SM_M_READSTATUS, // read status
+		SM_M_READID,     // read ID
+		SM_M_30,
+		SM_M_RANDOM_DATA_INPUT,
+		SM_M_RANDOM_DATA_OUTPUT
+	};
+
+	enum pointer_sm_mode_t
+	{
+		SM_PM_A, // accessing first 256-byte half of 512-byte data field
+		SM_PM_B, // accessing second 256-byte half of 512-byte data field
+		SM_PM_C  // accessing spare field
+	};
```

Comment:
Please give these explicit storage classes since you register them for save states (like `enum sm_mode_t : uint8_t { ... };` etc.).